src/highlights/highlight_group_store.ts

```diff
@@ -0,0 +1,157 @@
+import {
+    DecorationRangeBehavior,
+    Disposable,
+    TextEditorDecorationType,
+    ThemableDecorationRenderOptions,
+    ThemeColor,
+    window,
+} from "vscode";
+
+import { disposeAll } from "../utils";
+import { config } from "../config";
+import { VimHighlightUIAttributes } from "../eventBus";
+
+interface HighlightConfiguration {
+    /**
+     * Map specific highlight to use vscode decorator configuration
+     */
+    highlights: {
+        [key: string]: ThemableDecorationRenderOptions;
+    };
+}
+
+/**
+ * Convert VIM HL attributes to vscode text decoration attributes
+ * @param uiAttrs VIM UI attribute
+ * @param vimSpecialColor Vim special color
+ */
+function vimHighlightToVSCodeOptions(uiAttrs: VimHighlightUIAttributes): ThemableDecorationRenderOptions {
+    const options: ThemableDecorationRenderOptions = {};
+    // for absent color keys color should not be changed
+    if (uiAttrs.background !== undefined) {
+        options.backgroundColor = "#" + uiAttrs.background.toString(16).padStart(6, "0");
+    }
+    if (uiAttrs.foreground !== undefined) {
+        options.color = "#" + uiAttrs.foreground.toString(16).padStart(6, "0");
+    }
+
+    const specialColor = uiAttrs.special !== undefined ? "#" + uiAttrs.special.toString(16).padStart(6, "0") : "";
+
+    if (uiAttrs.reverse !== undefined) {
+        options.backgroundColor = new ThemeColor("editor.foreground");
+        options.color = new ThemeColor("editor.background");
+    }
+    if (uiAttrs.italic !== undefined) {
+        options.fontStyle = "italic";
+    }
+    if (uiAttrs.bold !== undefined) {
+        options.fontWeight = "bold";
+    }
+    if (uiAttrs.strikethrough !== undefined) {
+        options.textDecoration = "line-through solid";
+    }
+    if (uiAttrs.underline !== undefined) {
+        options.textDecoration = `underline ${specialColor} solid`;
+    }
+    if (uiAttrs.undercurl !== undefined) {
+        options.textDecoration = `underline ${specialColor} wavy`;
+    }
+    return options;
+}
+
+function normalizeThemeColor(color: string | ThemeColor | undefined): string | ThemeColor | undefined {
+    if (typeof color === "string" && color.startsWith("theme.")) {
+        color = new ThemeColor(color.slice(6));
+    }
+    return color;
+}
+
+function normalizeDecorationConfig(config: ThemableDecorationRenderOptions): ThemableDecorationRenderOptions {
+    const newConfig: ThemableDecorationRenderOptions = { ...config };
+    newConfig.backgroundColor = normalizeThemeColor(newConfig.backgroundColor);
+    newConfig.borderColor = normalizeThemeColor(newConfig.borderColor);
+    newConfig.color = normalizeThemeColor(newConfig.color);
+    newConfig.outlineColor = normalizeThemeColor(newConfig.outlineColor);
+    newConfig.overviewRulerColor = normalizeThemeColor(newConfig.overviewRulerColor);
+    return newConfig;
+}
+
+export class HighlightGroupStore implements Disposable {
```

Comment:
"store" seems unnecessary. Mentioning this so it doesn't become a pattern. 

```suggestion
export class HighlightGroups implements Disposable {
```

Reply:
My first idea was to name it like that as well, but using the plural form for a property feels a bit awkward. "store" is also used in Neovim's UI doc.